### ISSUE TITLE
Use pyspark.InheritableThread as the base thread class for threadpool

### ISSUE THREAD

**Issue by NivekNey** [joblib/joblib-spark#32]
**Context**

[InheritableThread](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.InheritableThread.html#pyspark.InheritableThread) is a thread class that pyspark provides to support sync between JVM and python threads.

The `ThreadPool` class [used in this package](https://github.com/joblib/joblib-spark/blob/91ad068ec82cdb45b765fcf5707b3dc299f37ee8/joblibspark/backend.py#L127) is python's original `multiprocessing.pool.ThreadPool`.

What goods/bads will it bring? Dunno, I'll start looking around.

**Proposal**

Without touching this package's source code, a user can already nicely make their own `ThreadPool` that uses their own `Thread` class. If this package were to adopt the pyspark Thread class, then the steps may be:

1. Create a custom [`DummyProcess`](https://github.com/python/cpython/blob/bce1418541a64a793960182772f985f64afbfa1a/Lib/multiprocessing/dummy/__init__.py#L34) that inherits `pyspark.InheritableThread`.
2. Create a custom [`ThreadPool`](https://github.com/python/cpython/blob/bce1418541a64a793960182772f985f64afbfa1a/Lib/multiprocessing/pool.py#L918) that uses the custom `DummyProcess`.
3. Use the custom `ThreadPool` in this package's `SparkDistributedBackend._get_pool` method.

**Comment by WeichenXu123** [comment]
Yes in future we will switch to it. InheritableThread can avoid memory leak in pyspark pin-thread mode (slightly leak not serious).
